Add an otelcol.auth.sigv4 component (#3200)
grafana/agent: docs/sources/flow/reference/components/otelcol.auth.sigv4.md

---
title: otelcol.auth.sigv4
---

# otelcol.auth.sigv4

`otelcol.auth.sigv4` exposes a `handler` that can be used by other `otelcol`
components to authenticate requests to AWS services using the AWS Signature Version 4 (SigV4) protocol. 

> See [here](https://docs.aws.amazon.com/general/latest/gr/signing-aws-api-requests.html) for more

Reviewer: Can you make this a named link so users know what they're clicking on?

The Google documentation style guide recommends against links like "here" in favor of being more descriptive: https://developers.google.com/style/link-text

components to authenticate requests to AWS services using the AWS Signature Version 4 (SigV4) protocol. 
See [here](https://docs.aws.amazon.com/general/latest/gr/signing-aws-api-requests.html) for more
information about SigV4.

> **NOTE**: `otelcol.auth.sigv4` is a wrapper over the upstream OpenTelemetry
> Collector `sigv4auth` extension. Bug reports or feature requests will be
> redirected to the upstream repository, if necessary.

Multiple `otelcol.auth.sigv4` components can be specified by giving them
different labels.

> **NOTE**: The Agent must have valid AWS credentials as used by the 
[AWS SDK for Go](https://aws.github.io/aws-sdk-go-v2/docs/configuring-sdk/#specifying-credentials).

## Usage

If the exporter URL does not contain enough information about the region and service, 
the minimum arguments are:
```river
otelcol.auth.sigv4 "LABEL" {
  region  = "REGION"
  service = "SERVICE"
}
```

In some cases the exporter URL does contain enough information about the region and service as 
described [here](#arguments). In that case there is no need to specify any arguments to this component:
```river
otelcol.auth.sigv4 "LABEL" {
}

> ```

Reviewer: To follow the recommendations documented in #3145, this section should be composed of nothing but a River code block with required arguments. 

We can document when region/service may be required in the Arguments block without providing any information here.

## Arguments

Name | Type | Description | Default | Required
---- | ---- | ----------- | ------- | --------

> `region` | `string` | The AWS region for the service you are exporting to for AWS SigV4. | "" | no
> `service` | `string` | The AWS service for AWS SigV4. | "" | no

Reviewer: Are these shorter descriptions still accurate?

```suggestion
`region` | `string` | The AWS region to sign with. | "" | no
`service` | `string` | The AWS service to sign with. | "" | no
```

Author: I think they are accurate, and I do think they sound better, thank you.

Both `region` and `service` are required for the component to work, but if they are left empty their values 

> may be inferred from the exporter URL based on 

Reviewer: They're not required, then :) I think this needs some rephrasing to talk about how they must be provided if the URL in the exporters don't follow a set of rules.

Author: Makes sense. What do you think about this instead:

> `region` and `service` only need to be specified if their values cannot be inferred from the exporter URL.
If they are left empty, their values may be inferred from the exporter URL based on the following rules:

Reviewer: Maybe flip it around?

> If `region` and `service` are left empty, their values are inferred from the URL of the exporter using this component using the following rules:
>
> (RULES)
>
> If none of the above rules apply, then `region` and `service` must be specified.

> [these](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/v0.63.0/extension/sigv4authextension/signingroundtripper.go#L121) rules:

Reviewer: IMO we should drop links to Go code since our readers might not be Go developers:

```suggestion
the following rules:
```

* If the exporter URL starts with `aps-workspaces` and `service` is empty, `service` will be set to `aps`.
* If the exporter URL starts with `search-` and `service` is empty, `service` will be set to `es`.
* If the exporter URL starts with either `aps-workspaces` or `search-` and `region` is empty, `region` .
will be set to the value between the first and second `.` character in the exporter URL.

> For concrete examples, see the [examples](#examples) section.

Reviewer: This shouldn't be necessary, readers will see on the sidebar that there is an examples section.

Author: Ok, I will delete this line.

> The AWS regions are listed [here](https://docs.aws.amazon.com/AmazonRDS/latest/UserGuide/Concepts.RegionsAndAvailabilityZones.html).

Reviewer: Avoid calling links "here", instead try something like:

```suggestion
A list of valid AWS regions can be found on Amazon's documentation for [Regions, Availability Zones, and Local Zones](https://docs.aws.amazon.com/AmazonRDS/latest/UserGuide/Concepts.RegionsAndAvailabilityZones.html).
```

This will then align with the Google developer documentation recommendations for [Link text](https://developers.google.com/style/link-text).

## Blocks

The following blocks are supported inside the definition of
`otelcol.auth.sigv4`:

Hierarchy | Block | Description | Required
--------- | ----- | ----------- | --------

> assume_role | [assume_role][] | Specifies the configuration needed to assume a role. | no

Reviewer: ```suggestion
assume_role | [assume_role][] | Configuration for assuming a role. | no
```

[assume_role]: #assume_role-block

### assume_role block

The `assume_role` block specifies the configuration needed to assume a role.

Name | Type | Description | Default | Required
---- | ---- | ----------- | ------- | --------
`arn` | `string` | The Amazon Resource Name (ARN) of a role to assume. | "" | no
`session_name` | `string` | The name of a role session. | "" | no
`sts_region` | `string` | The AWS region where STS is used to assume the configured role. | "" | no

Note that if a role is intended to be assumed and `sts_region` is not provided, then `sts_region`

> will default to the value for `region` if `region` is provided.

Author: Do you think this sounds better:

> If the `assume_role` block is specified in the config and `sts_region` is not set, then `sts_region` will default to the value for `region`

> will default to the value for `region` if `region` is provided.

Reviewer: Much better, thanks!

> `region` can be differentiated from `sts_region` to handle cross region authentication. 

Author: Do you mean "can be set to different values"?

> `region` can be differentiated from `sts_region` to handle cross region authentication. 

Reviewer: Yes, sorry

## Exported fields

The following fields are exported and can be referenced by other components:

Name | Type | Description
---- | ---- | -----------
`handler` | `capsule(otelcol.Handler)` | A value that other components can use to authenticate requests.

## Component health

`otelcol.auth.sigv4` is only reported as unhealthy if given an invalid
configuration.

## Debug information

`otelcol.auth.sigv4` does not expose any component-specific debug information.

## Examples

> In the example below the exporter endpoint starts with `aps-workspaces`. Hence `service` is inferred to be `aps`

Reviewer: Can you give each of these examples headers? That's not something mentioned (yet) in #3145 but I think it would help with the readability here.
